### DISCUSSION TITLE
Recommended environment information for Win32 and other. Feedback welcome.

### DISCUSSION THREAD

**Discussion by TAbdiukov** [upx/upx#741] [Q&A]
## What's the problem (or question)?

I couldn't find this information anywhere online, so my only sensible option seems to ask for this information here.

I want to create a virtual 32-bit environment for UPX pack/unpack operations of specifically 32-bit executables and libraries for an ongoing project. For that, I want to know what OS is recommended for it. For example, Windows 10, 7, 2000, XP, Server 2003, something else (perhaps ReactOS)?

Secondly, does it matter whether UPX executable is of the same architecture and platform as the target file?

Just for reference: I found a clue about my first question in [NEWS](https://github.com/upx/upx/blob/devel/NEWS), 

> Changes in 4.2.1 (01 Nov 2023):
>  * linux: /proc/self/exe now is optional
>  * windows: use SetFileTime to preserve sub-second file timestamps
>  * official Windows builds: revert activeCodePage change introduced in 4.2.0
>  * bug fixes - see https://github.com/upx/upx/milestone/14

**SetFileTime** is a function that requires [Windows XP/Server 2003](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-setfiletime)

Thank you for your time.

## What should have happened?
* Requested information made available

**Comment by markus-oberhumer** [comment]
### From the release notes at https://github.com/upx/upx/releases/latest :
- Note: all versions are functionally equivalent, i.e. each version can handle **all** executable formats, so you only need the file that runs on your host OS.

### Official release builds:

The official DOS version is cross-compiled with djgpp2 32-bit gcc. It should run on
- MSDOS 5 on Intel i486 or later
  - maybe even MSDOS 4 and/or Intel i386+FPU; but this is hard to test - you cannot trust emulators here and need real hardware to verify
 - Windows 95 until Windows 7 (or whenever Windows stopped supporting DOS programs)
 - FreeDOS https://www.freedos.org/
 - any other DOS compatible OS; feedback welcome
 
The official Win32 version is cross-compiled with Mingw-w64 32-bit gcc. It should run on
  - Windows XP SP3 and later
  - Windows64 XP SP2 and later
  - Wine https://www.winehq.org/
  - CPU: not sure if it runs on an Intel Pentium i586 CPU without MMX; feedback welcome

The official Win64 version is cross-compiled with Mingw-w64 64-bit gcc. It should run on
  - Windows64 XP SP2 and later
  - Wine https://www.winehq.org/

The official i386-linux version is statically linked against musl libc. It should run on
  - any Linux 2.6.x kernel (or maybe even 2.4.x kernel, please report back if you have a virtual machine)
  - CPU: any Intel Pentium i586 CPU (no MMX required)

I don't know about other versions like ReactOS or older Windows versions. If you guys have some older virtual machines please report back how upx-4.2.1 works there.

And I'd also be interested how UPX works on old DOS versions/really old i386 CPU, but as indicated above these tests must be done on real hardware.

### The future:

- at some point we might provide an official WASM/WASI build which then should run everywhere (even in the browser)
- this would also be interesting from a security POV

### Final note:

If this issue/discussion gives some interesting info I might convert this into a pinned issue.

And please always exactly describe which UPX version you are testing.

**Comment by markus-oberhumer** [comment]
[Empty placeholder comment]

**Comment by jreiser** [comment]
I suspect that upx itself is not the most troublesome piece for what I understand that you're doing.  Instead, I suggest paying more attention to the larger environment.  In particular: choose Windows 10 or a fully-updated Windows 7.  Windows 11 requires TrustedBoot, which can be troublesome in a VM.   Anything else (Windows 8 (Vista), WinXP, WinME, Win2000, ...) is so old that it can be hard to find personnel with experience.  And be sure to use the updated releases.  Initial release of Windows7 did not install onto GPT disk, and did not treat SSD nicely.
Or, just use Linux.
